core/iwasm/interpreter/wasm_interp_classic.c

```diff
@@ -1128,13 +1128,27 @@ wasm_interp_call_func_bytecode(WASMModuleInstance *module,
 {
     WASMMemoryInstance *memory = wasm_get_default_memory(module);
     uint8 *global_data = module->global_data;
+
 #if !defined(OS_ENABLE_HW_BOUND_CHECK)              \
     || WASM_CPU_SUPPORTS_UNALIGNED_ADDR_ACCESS == 0 \
     || WASM_ENABLE_BULK_MEMORY != 0
+#if WASM_ENABLE_SHARED_MEMORY != 0
+    WASMSharedMemNode *node =
+        wasm_module_get_shared_memory((WASMModuleCommon *)module->module);
+    if (node)
+        os_mutex_lock(&node->shared_mem_lock);
+#endif
     uint32 num_bytes_per_page = memory ? memory->num_bytes_per_page : 0;
     uint32 linear_mem_size =
         memory ? num_bytes_per_page * memory->cur_page_count : 0;
+#if WASM_ENABLE_SHARED_MEMORY != 0
+    if (node)
+        os_mutex_unlock(&node->shared_mem_lock);
 #endif
```

Comment:
This makes the coding style not so good, how about add a function to get the num_bytes_per_page and linear_mem_size:
```C
#if !defined(OS_ENABLE_HW_BOUND_CHECK)              \
    || WASM_CPU_SUPPORTS_UNALIGNED_ADDR_ACCESS == 0 \
    || WASM_ENABLE_BULK_MEMORY != 0
static uint32
get_num_bytes_per_page(WASMMemoryInstance *memory, void *node)
{
#if WASM_ENABLE_SHARED_MEMORY != 0
    if (node)
        os_mutex_lock(&((WASMSharedMemNode *)node)->shared_mem_lock);
#endif
   ...
}
static uint32
get_linear_memory_size(WASMMemoryInstance *memory, void *node)
{
    ...
}
#endif

And here:
#if WASM_ENABLE_SHARED_MEMORY != 0
    WASMSharedMemNode *node =
        wasm_module_get_shared_memory((WASMModuleCommon *)module->module);
#else
    void *node =NULL;
#endif
    uint32 num_bytes_per_page = get__num_bytes_per_page(memory, node);
    uint32 linear_mem_size = get_linear_mem_size(memory, node);
```

Reply:
The disadvantage is that we have to lock/unlock twice

Reply:
Yes, but usually we declare local variables at the beginning of `{ .. }`. Or you can declare the variables and get the node here, and set them after the variable declarations?

Reply:
> The disadvantage is that we have to lock/unlock twice

The num_bytes_per_page isn't changed in enlarge memory, no need to add lock for it. Only need to add lock for linear mem size.

Reply:
But the value still gets written (even if not changed)
https://github.com/bytecodealliance/wasm-micro-runtime/blob/9f0c4b63ac21c975aee7aae30236e306c8e49d13/core/iwasm/common/wasm_memory.c#L552 so that should still cause undefined behavior in case of a race condition. I'll apply the first suggestion.